status/condition_set.go

```diff
@@ -0,0 +1,295 @@
+// Inspired by https://github.com/knative/pkg/tree/97c7258e3a98b81459936bc7a29dc6a9540fa357/apis,
+// but we chose to diverge due to the unacceptably large dependency closure of knative/pkg.
+package status
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/samber/lo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ConditionTypes is an abstract collection of the possible ConditionType values
+// that a particular resource might expose.  It also holds the "root condition"
+// for that resource, which we define to be one of Ready or Succeeded depending
+// on whether it is a Living or Batch process respectively.
+type ConditionTypes struct {
+	root       ConditionType
+	dependents []ConditionType
+}
+
+// NewLivingConditions returns a ConditionTypes to hold the conditions for the
+// living resource. ConditionReady is used as the root condition.
+// The set of condition types provided are those of the terminal subconditions.
+func NewLivingConditions(d ...ConditionType) ConditionTypes {
+	return newConditionTypes(ConditionReady, d...)
+}
+
+// NewBatchConditions returns a ConditionTypes to hold the conditions for the
+// batch resource. ConditionSucceeded is used as the happy condition.
+// The set of condition types provided are those of the terminal subconditions.
+func NewBatchConditions(d ...ConditionType) ConditionTypes {
+	return newConditionTypes(ConditionSucceeded, d...)
+}
+
+func newConditionTypes(happy ConditionType, dependents ...ConditionType) ConditionTypes {
+	return ConditionTypes{
+		root:       happy,
+		dependents: lo.Reject(lo.Uniq(dependents), func(c ConditionType, _ int) bool { return c == happy }),
+	}
+}
+
+// ConditionSet provides methods for evaluating Conditions.
+// +k8s:deepcopy-gen=false
+type ConditionSet struct {
+	ConditionTypes
+	object Object
+}
+
+// For creates a ConditionSet from an object using the original
+// ConditionTypes as a reference. Status must be a pointer to a struct.
+func (r ConditionTypes) For(object Object) ConditionSet {
+	cs := ConditionSet{object: object, ConditionTypes: r}
+	// Set known conditions Unknown if not set.
+	for _, t := range append(r.dependents, r.root) {
+		if cs.Get(t) == nil {
+			cs.Set(Condition{
+				Type:     t,
+				Status:   lo.Ternary(cs.Root().IsTrue(), metav1.ConditionTrue, metav1.ConditionUnknown),
+				Severity: cs.severity(t),
+			})
+		}
+	}
+	return cs
+}
+
+// Root looks returns the root Condition
```

Comment:
An example of what this is to someone who is unfamiliar? My assumption is that this is the `Ready` condition in almost all cases